src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/ProgramAnalysis.java

```diff
@@ -1,31 +1,81 @@
 package at.ac.tuwien.kr.alpha.grounder.structure;
 
-import at.ac.tuwien.kr.alpha.common.Predicate;
-import at.ac.tuwien.kr.alpha.common.Program;
-import at.ac.tuwien.kr.alpha.grounder.NonGroundRule;
-
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Map;
 
+import at.ac.tuwien.kr.alpha.common.Predicate;
+import at.ac.tuwien.kr.alpha.common.Program;
+import at.ac.tuwien.kr.alpha.common.Rule;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.depgraph.DependencyGraph;
+import at.ac.tuwien.kr.alpha.grounder.FactIntervalEvaluator;
+import at.ac.tuwien.kr.alpha.grounder.Instance;
+import at.ac.tuwien.kr.alpha.grounder.NonGroundRule;
+
 /**
- * Copyright (c) 2017, the Alpha Team.
+ * Copyright (c) 2017-2019, the Alpha Team.
  */
 public class ProgramAnalysis {
 
-	private final Map<Predicate, HashSet<NonGroundRule>> predicateDefiningRules;
+	private final Map<Predicate, HashSet<NonGroundRule>> predicateDefiningRules = new HashMap<>();
+	private final Map<Predicate, LinkedHashSet<Instance>> programFacts = new LinkedHashMap<>();
+	private final Map<Integer, NonGroundRule> nonGroundRules = new HashMap<>();
+	private final DependencyGraph dependencyGraph;
 
 	public ProgramAnalysis(Program program) {
-		predicateDefiningRules = new HashMap<>();
+		this.initializeFacts(program);
+		this.initializeRules(program);
+		this.dependencyGraph = DependencyGraph.buildDependencyGraph(this.nonGroundRules);
+	}
+
+	private void initializeFacts(Program program) {
+		List<Instance> tmpInstances;
+		Predicate tmpPredicate;
+		for (Atom fact : program.getFacts()) {
+			tmpInstances = FactIntervalEvaluator.constructFactInstances(fact);
+			tmpPredicate = fact.getPredicate();
```

Comment:
I am not happy with the names here: `tmpInstances` are the interval-expanded instances/facts while the `tmpPredicate` is the predicate of all those facts.